hpc_benchmark.sli is now invariant under different MPI-process/thread splits with fixed VP number

stinebuu opened the pull request:
This PR fixes #247.

I have added a new example, `hpc_benchmark_invariant.sli`, a mpi/thread-split invariant version of `hpc_benchmark.sli`. To make it invariant under mpi/thread splits, we
- record from a fixed set of neurons (the Nrec first exitatory), instead of a fixed number of local neurons
- randomize the membrane potential based on the virtual processes, not the mpi processes.

With this new version, we get the following average rates (also averaged over processes). I include the rates from hpc_bencmark as well for comparison, scale = 1.0.

| Num_vps | Split | hpc_benchmark_invariant | hpc_benchmark |
|:------------:|:------:|:-----------------------------------:|:---------------------:|
|        1       | 1 mpi proc., 1 thread   |  11.832  |        11.832         |
|        4       | 1 mpi proc., 4 threads |    13.5    |        11.812         |
|        4       | 2 mpi proc., 2 threads |    13.5    |         29.06          |
|        4       | 4 mpi proc., 1 thread   |    13.5    |         54.58          |

jakobj commented:
sounds good, but could you elaborate why the rates increases so much for the old hpc_benchmark? this is hopefully just a measurement artifact?

stinebuu commented:
@jakobj Yes, it has to do with the neurons we record from. If I use the record strategy from hpc_benchmark_invariant in the hpc_benchmark, that is, always record from the Nrec first neurons and not let recording be based on the mpi processes, the increase in rate is much lower. Then the rate is 11.812 for 1 mpi proc. 4 threads, 14.68 for 2 mpi procs. 2 threads and 13.49 for 4 mpi procs. 1 thread.

heplesser commented:
@stinebuu Nice, but I would suggest to also correct the firing rate calculation in the original `hpc_benchmark.sli`, so that one gets sensible results.

Should we in addition to the average rate also report 
- the total spike count
- the number of nodes
- the number of connections?

jakobj commented on examples/nest/hpc_benchmark_invariant.sli:
```diff
@@ -0,0 +1,512 @@
+    E_neurons size Nrec lt
+    {
+      M_ERROR (BuildNetwork)
+      (Networks is smaller than number of neurons to record from!) message
```
since your only checking the size of the E population, the error message should reflect this

jakobj commented on examples/nest/hpc_benchmark_invariant.sli:
```diff
@@ -0,0 +1,512 @@
+
+  tic
+
+  Prepare
```
why use prepare, run, cleanup instead of simulate? do the advantages outweigh the change wrt the hpc_benchmark?

stinebuu replied:
I think I must have been inspired by another benchmark. Tested quickly now, and it does not seem to have an effect, will change to Simulate.

stinebuu commented:
@heplesser In #247 it was decided that we should keep the original `hpc_benchmark.sli` as constant as possible as it is used in many publications. If I change the firing rate calculations the benchmark will change a bit, should I still change the benchmark? I am also wondering if we would then really need the invariant version? The difference between the invariant version and the original hpc_benchmark is basically just the firing rate calculations and the detector scheme, so if we change one we might as well change both.

I'll add the total spike count, number of nodes and number of connections to `hpc_benchmark.sli`, it is already reported in `hpc_benchmark_invariant.sli`. :)

heplesser commented:
@stinebuu I agree that the changes are so small that we probably can just update `hpc_benchmark.sli`, but I would be interested in @suku248's take on this.

I also think we should move the definitions of `/max_rank_cout` and `/max_rank_log`, which are now hidden in the definition of the `/logger` function forward to the parameter section to make them visible and accessible to the user.

stinebuu commented:
@jakobj @jarsi @heplesser I have now renamed the original hpc_benchmark to be `hpc_benchmark_2_16.sli` and called my invariant one for `hpc_benchmark.sli`. I have added a comment in `hpc_benchmark_2_16.sli` about it being the original one, so if you want to reproduce your old results, you should use that one.
